mlir/lib/Quantum/Transforms/DisentangleCNOT.cpp

```diff
@@ -0,0 +1,144 @@
+// Copyright 2024 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This algorithm is taken from https://arxiv.org/pdf/2012.07711, table 1
+
+#define DEBUG_TYPE "disentanglecnot"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/Support/Debug.h"
+
+#include "Catalyst/IR/CatalystDialect.h"
+#include "Quantum/IR/QuantumOps.h"
+
+#include "PropagateSimpleStatesAnalysis.hpp"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst;
+
+namespace catalyst {
+#define GEN_PASS_DEF_DISENTANGLECNOTPASS
+#define GEN_PASS_DECL_DISENTANGLECNOTPASS
+#include "Quantum/Transforms/Passes.h.inc"
+
+struct DisentangleCNOTPass : public impl::DisentangleCNOTPassBase<DisentangleCNOTPass> {
+    using impl::DisentangleCNOTPassBase<DisentangleCNOTPass>::DisentangleCNOTPassBase;
+
+    bool canScheduleOn(RegisteredOperationName opInfo) const override
+    {
+        return opInfo.hasInterface<FunctionOpInterface>();
+    }
+
+    void runOnOperation() override
+    {
+        func::FuncOp func = cast<func::FuncOp>(getOperation());
```

Comment:
These two lines don't quite seem to match, unless the `FunctionOpInterface` is guaranteed to be exclusive to the `FuncOp`.

Reply:
Good point! Changed the variable to `FunctionOpInterface` instead since [that is the recommendation of the mlir docs](https://mlir.llvm.org/docs/PassManagement/#operation-pass-static-schedule-filtering) (https://github.com/PennyLaneAI/catalyst/pull/1154/commits/2b6b9dd26864dc4727b244b704b88381435dc6d9)